### PR TITLE
imdsv2 support

### PR DESCRIPTION
**Pull request by bwagner5**
Issue #, if available:
https://github.com/aws/aws-node-termination-handler/issues/22

Description of changes:
Support IMDSv2 w/ an IMDSv1 fallback. There is no configuration since imdsv2 should always be favored, but will fallback to imdsv1 if there is an issue or if it's blocked due to a container hop if not configured with a TTL of 2. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmd/node-termination-handler.go [hunk 1]
```diff
@@ -24,16 +24,18 @@ import (
...
 	"github.com/aws/aws-node-termination-handler/pkg/draineventstore"
+	"github.com/aws/aws-node-termination-handler/pkg/ec2metadata"
 	"github.com/aws/aws-node-termination-handler/pkg/node"
 	"github.com/aws/aws-node-termination-handler/pkg/webhook"
 )
 
 const (
 	scheduledMaintenance = "Scheduled Maintenance"
 	spotITN              = "Spot ITN"
+	metadataTries        = 3
```
Comment (mattrandallbecker) [inline cmd/node-termination-handler.go]: how do you feel about moving this outside this file? feels like its part of configuration unrelated to which monitoring functions we are tracking

File: config/helm/ec2-metadata-test-proxy/values.yaml [hunk 2]
```diff
@@ -7,6 +7,7 @@ ec2MetadataTestProxy:
     scheduledEventStatus: active
     enableScheduledMaintenanceEvents: false
     enableSpotITN: true
+    enableIMDSV2: false
```
Comment (mattrandallbecker) [inline config/helm/ec2-metadata-test-proxy/values.yaml]: why is this a value at all? it doesn't seem used in the code and even if it's set to false we seemingly try v2 in each request unless im missing something

File: pkg/drainevent/scheduled-event.go [hunk 3]
```diff
@@ -56,17 +54,11 @@ func MonitorForScheduledEvents(drainChan chan<- DrainEvent, cancelChan chan<- Dr
 }
 
 // checkForScheduledEvents Checks EC2 instance metadata for a scheduled event requiring a node drain
-func checkForScheduledEvents(metadataURL string) ([]DrainEvent, error) {
-	resp, err := ec2metadata.RequestMetadata(metadataURL, ec2metadata.ScheduledEventPath)
+func checkForScheduledEvents(imds *ec2metadata.EC2MetadataService) ([]DrainEvent, error) {
+	scheduledEvents, err := imds.GetScheduledMaintenanceEvents()
```
Comment (mattrandallbecker) [inline pkg/drainevent/scheduled-event.go]: this is so much better

File: pkg/ec2metadata/ec2metadata.go [hunk 4]
```diff
@@ -67,12 +89,135 @@ type InstanceAction struct {
...
+		tries:       tries,
+		httpClient:  http.Client{},
+	}
+}
+
+// GetScheduledMaintenanceEvents retrieves EC2 scheduled maintenance events from imds
+func (e *EC2MetadataService) GetScheduledMaintenanceEvents() ([]ScheduledEventDetail, error) {
+	resp, err := e.Request(ScheduledEventPath)
+	if resp != nil && (resp.StatusCode < 200 || resp.StatusCode >= 300) {
+		return nil, fmt.Errorf("Unable to request metadata, received http status code: %d", resp.StatusCode)
```
Comment (mattrandallbecker) [inline pkg/ec2metadata/ec2metadata.go]: nit, we were able to request metadata, we just didnt like the response

File: test/k8s-local-cluster-test/run-test [hunk 5]
```diff
@@ -185,8 +193,11 @@ export EC2_METADATA_DOCKER_REPO=$(echo $EC2_METADATA_DOCKER_IMG | cut -d':' -f1)
 export EC2_METADATA_DOCKER_TAG=$(echo $EC2_METADATA_DOCKER_IMG | cut -d':' -f2)
 ###
 
+i=0
 for assert_script in $ASSERTION_SCRIPTS; do
   reset_cluster
+  export IMDS_PORT=$(expr $i + 1338)
```
Comment (mattrandallbecker) [inline test/k8s-local-cluster-test/run-test]: much smarter port collision mechanism, thank you

File: pkg/drainevent/scheduled-event.go (hunk 3 above)
Comment (bwagner5) [reply pkg/drainevent/scheduled-event.go]: yep, that's a good idea

File: cmd/node-termination-handler.go (hunk 1 above)
Comment (bwagner5) [reply cmd/node-termination-handler.go]: yep, that's a good idea

**Pull request merged**